### PR TITLE
more flexible options for building/running scripts

### PR DESCRIPTION
**Pull request by krancour**
This change allows users to provide more specific instructions on how to build and run their script, but gracefully degrades to sensible default strategies in the absence of those specific instructions.

cc @itowlson. lmk if this improves things for you.

### CODE REVIEW

**Review by itowlson — COMMENTED** [review]
This behaviour looks very promising - thanks!  One thing I would suggest is to refactor the code to reduce the nesting.  A possible approach would be to define `BuildStrategy` and `RunStrategy` interfaces, with appropriate implementations for yarn, npm, raw TS and raw JS.  Then do a detection pass up top to figure out which strategies to use and then just call them.  E.g.

```
const packager = fs.existsSync('yarn.lock') ? 'yarn' : 'npm';
let buildStratey: BuildStrategy;
if (packageJSON && packageJSON.scripts && packageJSON.scripts.build) {
  buildStrategy = (packager === 'yarn') ? new BuildYarn() : new BuildNPM();
} else if (fs.existsSync('tsconfig.json')) {
  buildStrategy = new BuildTypescriptProject();
} else if (fs.existsSync('brigade.ts')) {
  buildStrategy = new BuildTypescriptLooseFile();
} else {
  buildStrategy = new DontBuild();
}

buildStrategy.build();
```

For me this makes the decision tree much more discoverable, and isolates out the logic of each strategy into a smaller component.  But the downside is it creates a lot more of those small components and the code probably becomes longer overall.

Either way this looks like a great step to support scripts that make use of third party dependencies or which merit splitting up into multiple files.  Thanks!

**Comment by krancour** [comment]
@itowlson thanks for looking at this. I'll be cleaning it up a bit today.

File: v2/worker/src/prestart.ts [hunk 1]
```diff
@@ -60,42 +75,103 @@ if (fs.existsSync(packageJSONPath)) {
...
+        cwd: configFilesPath,
+        stdio: debug ? "inherit" : undefined
+      })
+    } catch(e) {
+      throw new Error(`error executing yarn build:\n\n${e.output}`)
+    }
+  } else {
+    logger.debug("running build script with npm")
+    try {
+      execFileSync("npm", ["run-script", "build"], {
```
[the remark this answers is not in the document]
Comment (krancour) [reply v2/worker/src/prestart.ts]: Good call. I've now done so.

File: v2/worker/src/prestart.ts [hunk 2]
```diff
@@ -60,42 +75,103 @@ if (fs.existsSync(packageJSONPath)) {
...
+        stdio: "inherit"
+      })
+    } catch(e) {
+      throw new Error(`error executing yarn run run:\n\n${e.output}`)
+    }
+  } else {
+    logger.debug("running script with npm")
+    try {
+      execFileSync("npm", ["run-script", "run"], {
+        cwd: configFilesPath,
```
[the remark this answers is not in the document]
Comment (krancour) [reply v2/worker/src/prestart.ts]: Another good call.

File: v2/worker/src/prestart.ts [hunk 3]
```diff
@@ -31,13 +31,28 @@ if (event.worker.defaultConfigFiles) {
   }
 }
 
+// Figure out whether we should use npm or yarn as the package manager. If we
+// find yarn.json, we use yarn. Otherwise, we use npm.
```
[the remark this answers is not in the document]
Comment (krancour) [reply v2/worker/src/prestart.ts]: Good eye. Fixed.

**Comment by krancour** [comment]
> A possible approach would be to define BuildStrategy and RunStrategy interfaces, with appropriate implementations for yarn, npm, raw TS and raw JS...

I agree that the logic could be made less nested and more readable. The interfaces might be overkill. I've split the different and created small functions for things like `yarnInstall()`, `npmInstall()`, various methods of building, running, and so forth.

I think it reads much cleaner now without getting too crazy.

**Comment by krancour** [comment]
I accidentally closed this by deleting the branch. Fixing that now.

**Comment by krancour** [comment]
Apparently, after deleting a branch, I just need to make a new PR.

**Pull request closed without merging**